research/conditional/utils/layer_manager.py

```diff
@@ -32,16 +32,31 @@ def __init__(
 
     def _register_layers(self, model):
         for name, layer in model.named_modules():
-            if name.endswith("feedforward"):
-                pattern = r"block_(\d+)"
-                match = re.search(pattern, name)
-                if match:
-                    block_name = match.group()
-                else:
-                    raise Exception(
-                        f"The expected pattern {pattern} was not found in name: {name}. The naming convention of model layers is not as expected."
-                    )
+            suffix = name.split(".")[-1]
+            if "residual" in suffix:
+                block_name = self.extract_block_name(name)
+                subblock_type = (
+                    "feedforward"
+                    if hasattr(layer.layer, "feedforward")
+                    else "attention"
+                )
+                self._layers.append((f"{block_name}_{subblock_type}_residual", layer))
+                print(f"Registered layer {block_name}_{subblock_type}_residual")
+            elif "feedforward" in suffix:
+                block_name = self.extract_block_name(name)
                 self._layers.append((block_name, layer))
+                print(f"Registered layer {block_name}")
```

Comment:
Spent a couple of minutes trying to understand that, but I lack background in our naming convention and logging. I am especially lost with `hasattr(layer.layer, "feedforward")`. 
@Simontwice can you either:
a) give a thorough explanation of this part of code
b) find a reviewer familiar with logging layers, e.g. @kuba-krj

Reply:
I think now it's much more self-explanatory, pls take a look @maciek-pioro